penguin/tools/runtime.py

```
@@ -0,0 +1,457 @@
+        "action": tool_result.name,
+        "result": tool_result.output,
+        "status": tool_result.status,
+        "tool_call_id": tool_result.call_id,
+        "output_hash": tool_result.output_hash,
+        "metadata": structured_output,
+        "tool_arguments": structured_output.get("tool_arguments")
+        or structured_output.get("arguments"),
+    }
+
+
+def _normalize_tool_loop_input(action_result: Any) -> dict[str, Any]:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `action_result`

(ANN401)